### DISCUSSION TITLE
XCOM push fails if a datetime object is in the return value

### DISCUSSION THREAD

**Discussion by ferruzzi** [apache/airflow#24881] [General]
### Apache Airflow version

2.3.2 (latest released)

### What happened

If you have an Operator returning a dict value such as from an API call, and that dict contains datetime objects, XCOM will fail with the message "datetime object not serializable"

For example, the SageMaker operators return a dict which includes the timestamp that the object was created, see https://github.com/apache/airflow/pull/23628

### What you think should happen instead

The suggestion from @uranusjr was to use 'AirflowJsonEncoder' directly in XCom’s 'serialize' function.

### Operating System

not relevant

### Deployment

Other

### Anything else

I have this on my backlog to fix, but figure if I put it out there, someone else may get to it sooner.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I think we should treat it  very carefully. Changing the way how we store XCom is not something that we should do on global level - and certainly not something that we should enable "globally" for everything. The problem is that the way how we store XCom is not an interanl detail of Airflow, but it is also displayed in the UI and are part of the public API. 

https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_xcom_entry

While it is not specify what the "string" is, there are likely already many users relying on certain form of it. 

So any change that would affect all operator's way of XCom serialization form have to not only be displayable in the UI, but it also should produce the same form for all the current operators. So I would be really cautious about it.

I think possibly a better approach would be to use AirflowJSon Serializer by the SageMaker Operator. That would be  much less intrusive way (only affecting the operator) and we can also make it a "recommended" solution to any other operator that wants to return non-serializable fields.  This way we have a good "solution" we can propose "anyone", but we do not change "everyeone's" behaviour. I think that's good enough or actually even "good". 

Let me convert it into a discussion. It certainly is not a "bug"

**Comment by ferruzzi** [comment]
Sagemaker can't be the only operators returning a date, can it?  How do other operators with datetime objects handle it if they are not all failing?

**Reply by potiuk** [reply]
returning `datetime.isoformat()` maybe?

**Reply by potiuk** [reply]
The root of the problem is that what happens is that you try to return a native "SageMaker" return value as Xcom. and this is generally speaking not a good idea if the object was not created with being serializable in the first place. Xcom by definition is and should be serializable.

**Reply by potiuk** [reply]
And again - I have nothing against SageMaker using AirflowJsonSerializer - every operator is free to define what their XComs will be and how they are formatted. I just don't see Airflow to try to attempt to parse more complex format on their own. Just having a requirement "XCom must be serializable object" and providing a tool that allows to serialize more complex data structures is (IMHO) more than enough. There is virtually no benefit of making it a "default" action for all operators to use "AirlfowJSonSerializer". 

 IMHO, if any operator can make a decision to do it on their own when they want to pass potentially non-serializable data. No harm done.

**Reply by shubham22** [reply]
@potiuk I'm trying to understand how using `AirflowJsonEncoder` (assuming you meant this with `AirflowJsonSerializer`) in SageMaker would be better from what @ferruzzi originally proposed in the PR https://github.com/apache/airflow/pull/23628 
I think both are localized to an operator and can be recommended generally.

**Reply by potiuk** [reply]
> @potiuk I'm trying to understand how using `AirflowJsonEncoder` (assuming you meant this with `AirflowJsonSerializer`) in SageMaker would be better from what @ferruzzi originally proposed in the PR #23628 I think both are localized to an operator and can be recommended generally.

I never said it is better. Read what I wrote in detail.  I just said that forcing new kind of serialization for all xcoms returned by all operators is bad idea. I am perfectly fine if SageMaker does not use `AirflowJsinSerizalizer` but custom way of serializing. This is perfectly fine for me.

But I am against "blanket" changing of serialization mechanism for all operators, because it would be backwards-incompatible. No more, no less. This discussion is about it I beleve?

**Reply by potiuk** [reply]
Just to say it differently - operators IMHO should be free to use how they make their XCom "serializable". AirlfowJsonSerializer is something that we could offer as a solution. But "forcing" a solution to be AirlfowJsonSerializer is pretty wrong IMHO. 

BTW. I think it's just a question of interpretatiuon of written communication. Where did you get from that "AirflowSerializerJson" is better than the original proposal of @ferruzzi  ? I am genuinly curious on how you got to that conclusion :).  I am trying to reflect back on things I do and say and I am curious to see where I made a mistake to mislead you to think like that?

**Reply by shubham22** [reply]
Got it! That's what I thought as well, but wanted to confirm again. Thank you for the explanation :) 

> I am against "blanket" changing of serialization mechanism for all operators, because it would be backwards-incompatible.

I am with you on this and totally agree that we can't change the overall behavior based on 1 instance. Any changes to XCom serialize would require further deep dive.

**Reply by potiuk** [reply]
BTW. I re-read it - the "possibly better" was NOT better than original proposal but better than "blanket forcing of serialization using AirlfowJsonSerialization". This is what I meant. I was referring to "current" issue proposal - not #23628 one. This is a PR, that was  leading to creating this one - but I was commenting on the proposal here - not there.

This is what it is better than:

> ## What you think should happen instead
> The suggestion from @uranusjr was to use 'AirflowJsonEncoder' directly in XCom’s 'serialize' function.

**Reply by shubham22** [reply]
> Where did you get from that "AirflowSerializerJson" is better than the original proposal

This was probably _my_ interpretation only as I did want to double confirm if there is something we can improve in the original proposal on https://github.com/apache/airflow/pull/23628, not here (I see how this got so convoluted, sorry about it). Again, not an issue with your communication style, probably a side-effect of me mixing up this discussion and the SageMaker PR.

**Reply by potiuk** [reply]
Well, actualy I see where I could improve :). That's good example of - if you are referring to something, quote it rather than say "it" because "it" can be understood differently.

**Reply by shubham22** [reply]
True. Good point! Applies to me as well :D

**Comment by funes79** [comment]
Hi! I think I bumped into the same problem, but this time with SnowflakeOperator. Its not clear for me how can I pass a custom function or this AirflowJson Serialization object to the operator. 
```
SnowflakeOperator(task_id='sqlop', snowflake_conn_id='SF', sql='select MAX(event_date) from MYDB.MYSCHEMA.TBL_EVENTS;', results_serializer=<?> )
```

```
  File "/home/ttt/.env/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 2290, in xcom_push
    XCom.set(
    ...
    
     CLASSNAME: o.__module__ + "." + o.__class__.__qualname__,
AttributeError: 'datetime.datetime' object has no attribute '__module__'
```

I checked the discussed SageMaker operator, and it is different than the Snowflake, the code itself has a serialization. 
So to support that do I need to change the Operator itself? Like here? https://github.com/apache/airflow/blob/main/airflow/providers/snowflake/operators/snowflake.py#L118

Thanks

**Reply by lucas3900** [reply]
I also ran into this problem, with same error. One possible solution I thought of was bypassing the data parsing done by the snowflake connector like here: https://docs.snowflake.com/en/user-guide/python-connector-example#label-python-connector-bypass-data-conversion

But I don't see any way to pass this via the `SnowflakeOperator`

**Reply by glek3327** [reply]
add this inside SnowFlakeOperator :
```python
  do_xcom_push=False,
``` 
that parameter's default value is True, which causes task attempt pushing the query result into xcom.

ref. https://github.com/apache/airflow/blob/main/airflow/models/baseoperator.py#L790
